### DISCUSSION TITLE
Specific Round Not Calling round_freeze_end

### DISCUSSION THREAD

**Discussion by c9win** [saul/demofile#271] [Questions] [answered]
On round 21 of Inferno on this match (https://www.hltv.org/matches/2343136/big-vs-heroic-dreamhack-open-summer-2020-europe), it doesn't call the gameEvents round_freeze_end and I am not sure if it is calling any events in the round either, because, after round_start, when I capture the first event after round_start of round 21, it shows demoFile.entities.gameRules.roundNumber as 22 and seems to have skipped over the round. So, something weird seems to be happening. Watched the vod and nothing weird happens there.

I have done many other demos with no issues. So, I am not sure what is weird about that round, but, thought I would point it out to see if it replicates for you and if its a random bug.

Edit: Also happened again in Round 2 of Mirage (map 3) here: https://www.hltv.org/matches/2343140/big-vs-og-dreamhack-open-summer-2020-europe. Maps 1 and 2 were fine.

**Comment by Fhoerth** [comment]
I can confirm this. I'm trying to calculate how much time passes for an event since freeze time ends and I'm not able to do this for those rounds where `round_freeze_end` event is not called.
I tried to solve this with a workaround by registering the first player's footstep after a round starts but is not accurate.

**Comment by Fhoerth** [comment]
As an alternative, this works for me:

```
demoFile.gameEvents.on('round_start', () => {
  const frameTime = 1000 / demoFile.tickRate;
  const freezeTime = parseInt(demoFile.conVars.vars.get('mp_freezetime'), 10);
  const freezeTimeEndTick = demoFile.currentTick + Math.round((freezeTime * 1000) / frameTime);
  
  ...
});
```

By the way: I can confirm at least in the Round 2 of Mirage (big vs og), all events are being called except for `round_freeze_end`.

**Comment by c9win** [comment]
@Fhoerth Thanks for posting a workaround. Do you know if that work if the team calls timeout or there is a pause?

**Comment by Fhoerth** [comment]
@c9win Hi! The previous workaround doesn't work with timeouts, nor pause. That's right.

I came with another workaround:
```
  demoFile.gameEvents.on('cs_round_start_beep', () => {
    const freezeTimeEndTick = demoFile.currentTick + demoFile.tickRate;
  });
```
Let me explain what it does:

This event is called when the "beep" sounds when timeout is ending. The last beep sounds when 1 second of timeout is left.
Take note that `demoFile.tickRate` is the number of ticks for 1 second.

This should work unless a timeout is fired between the second 1 and 0 of the freezetime.

Let me know if you find another solution.

** Edit: Would be very complicated and ugly, but I think it should be possible to add a listener to `mp_pause_match` `mp_unpause_match` commands and modify the `freezeTimeEndTick` value if pause occurs between 1 and 0 of `freeze time`.
Maybe the same can be done for timeouts (any hint?).

**Comment by Fhoerth** [comment]
By the way, I don't think this is an issue related to `demofile`.
I don't have enough knowledge to parse the *.dem using https://github.com/ValveSoftware/csgo-demoinfo and check if the event exists, but I'm pretty sure that this is an issue with replays itself.

**Comment by saul** [comment]
Hi all

I can't try this as I'm not at a computer for the holiday season, but try watching the `DT_CSGameRules.m_bFreezePeriod` game rules net prop. 

Let me know how you get on

**Comment by Fhoerth** [comment]
Hey, seems that the table is being updated:
```
  demoFile.entities.on('change', e => {
    if (e.tableName === 'DT_CSGameRules' && e.varName === 'm_bFreezePeriod') {
      console.log('m_bFreezePeriod changed to, e.newValue);
    }
  });
```

**Comment by Fhoerth** [comment, marked as answer]
So, up to now, the best workaround seems to be:

```
  demoFile.entities.on('change', ({ tableName, varName, newValue }) => {
    if (tableName === 'DT_CSGameRules' && varName === 'm_bFreezePeriod' && !newValue) {
      demoFile.gameEvents.emit('round_freeze_end', {});
    }
  });
```

In my case, I don't really care that `round_freeze_end` is fired twice for most of the rounds since I only use that event to set up some variables.

**Comment by saul** [comment]
Closing as answered - see the solution above.
